File: backend/utils/fingerprint.ts
import crypto from 'crypto';

export interface ScannedData {
  name?: string;
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  ingredients: string[];
  userProfile?: {
    language?: string;
    goals?: string[];
    activityLevel?: string;
    explanationDepth?: string;
  };
}

export const generateHash = (data: ScannedData): string => {
  const normalizedIngredients = [...data.ingredients]
    .map(i => i.toLowerCase().trim())
    .sort() // Alphabetize so order doesn't break the cache
    .join(',');

  const profile = data.userProfile || {};
  const language = (profile.language || 'English').toLowerCase();
  
  // Sort goals alphabetically so ["Fat Loss", "Muscle"] hashes the same as ["Muscle", "Fat Loss"]
  const goals = [...(profile.goals || [])]
    .sort()
    .join(',')
    .toLowerCase();
    
  const activity = (profile.activityLevel || 'Active').toLowerCase();
  const depth = (profile.explanationDepth || 'Give me some context').toLowerCase();

  const rawString = `${data.calories}-${data.protein_g}-${data.carbs_g}-${data.fat_g}-${normalizedIngredients}-${language}-${goals}-${activity}-${depth}`;

  return crypto.createHash('sha256').update(rawString).digest('hex');
};

File: backend-py/app/services/fingerprint.py
import hashlib

from app.schemas import ScannedData


def _num(value: float) -> str:
    # Match JavaScript's number-to-string (150 -> "150", not "150.0") so keys
    # stay identical to the ones the Next.js backend wrote.
    return str(int(value)) if value == int(value) else repr(value)


def generate_hash(data: ScannedData) -> str:
    ingredients = ",".join(sorted(i.strip().lower() for i in data.ingredients))

    profile = data.userProfile
    language = ((profile and profile.language) or "English").lower()
    # Sorted so ["Fat Loss", "Muscle"] and ["Muscle", "Fat Loss"] share a key
    goals = ",".join(sorted((profile and profile.goals) or [])).lower()
    activity = ((profile and profile.activityLevel) or "Active").lower()
    depth = ((profile and profile.explanationDepth) or "Give me some context").lower()

    raw = (
        f"{_num(data.calories)}-{_num(data.protein_g)}-{_num(data.carbs_g)}-{_num(data.fat_g)}"
        f"-{ingredients}-{language}-{goals}-{activity}-{depth}"
    )
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()
